x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecyclePolicyUsageCalculator.java

```diff
@@ -0,0 +1,105 @@
+    /** Whether {@link #calculateUsage} will be called multiple times or not. */
+    private final boolean willIterate;
+    /** A map from policy name to list of data streams that use that policy. */
+    private final Map<String, List<String>> policyToDataStream;
+    /** A map from composable template name to the policy name it uses (or null) */
+    private final Map<String, String> templateToPolicy;
+
+    public LifecyclePolicyUsageCalculator(
+        final IndexNameExpressionResolver indexNameExpressionResolver,
+        final ClusterState state,
+        List<String> names
+    ) {
+        this.state = state;
+        this.willIterate = names.size() > 1 || Regex.isSimpleMatchPattern(names.get(0));
+
+        var allDataStreams = indexNameExpressionResolver.dataStreamNames(state, IndicesOptions.LENIENT_EXPAND_OPEN_CLOSED_HIDDEN);
+        // Sort all templates by descending priority. That way, findV2Template can exit on the first found template.
+        var indexTemplates = new ArrayList<>(state.metadata().templatesV2().entrySet());
+        CollectionUtil.timSort(indexTemplates, Comparator.comparing(entry -> entry.getValue().priorityOrZero(), Comparator.reverseOrder()));
+
+        // Build the maps that will be used for the usage calculation later on.
+        IndexLifecycleMetadata metadata = state.metadata().custom(IndexLifecycleMetadata.TYPE);
+        policyToDataStream = new HashMap<>(Regex.isSimpleMatchPattern(names.get(0)) ? metadata.getPolicyMetadatas().size() : names.size());
+        templateToPolicy = new HashMap<>(indexTemplates.size());
+        for (String dataStream : allDataStreams) {
+            String indexTemplate = MetadataIndexTemplateService.findV2Template(state.metadata(), indexTemplates, dataStream, false, true);
+            if (indexTemplate == null) {
+                continue;
+            }
+            Settings settings = MetadataIndexTemplateService.resolveSettings(state.metadata(), indexTemplate);
+            var policyName = LifecycleSettings.LIFECYCLE_NAME_SETTING.get(settings);
+            if (names.stream().noneMatch(name -> Regex.simpleMatch(name, policyName))) {
+                // If a template's policy doesn't match any of the supplied names, we can skip it later on.
+                templateToPolicy.put(indexTemplate, null);
+                continue;
+            }
+            templateToPolicy.put(indexTemplate, policyName);
+            policyToDataStream.computeIfAbsent(policyName, k -> new ArrayList<>()).add(dataStream);
+        }
+    }
+
+    /**
+     * Calculate the indices, data streams, and composable templates that use the given policy.
+     */
+    public ItemUsage calculateUsage(String policyName) {
+        List<String> indices = new ArrayList<>();
+        for (IndexMetadata indexMetadata : state.metadata().indices().values()) {
+            if (policyName.equals(indexMetadata.getLifecyclePolicyName())) {
+                indices.add(indexMetadata.getIndex().getName());
```

Comment:
I _could_ pre-compute a map of policy to indices as well, but I'm not sure the memory vs. speed trade-off is worth it there.

Reply:
On second thought, since we have to build this list for every policy anyway, pre-computing these lists and putting them in a map shouldn't be too much additional memory overhead. I'll wait till someone has done a first review before making more changes (in case my whole approach is off).

Reply:
And the same goes for the composable templates of course.

Reply:
Let's think about this. You only need to keep a cache of what is necessary. Right? 

For example:
**Composable templates:**
Cache the resolved templates that match our policies and their index patterns: `template -> policy`, probably you can also store `policy -> templates`.

**Data streams**
Go over the data streams like you already do, find the template for each data stream and check the cache `template -> policy`. You do not need to keep track of nulls anymore, since you have collected the relevant templates, if the template is not in your cache then the data stream shouldn't be either. So you create the `policy -> data stream`.

**Indices**
I do not see an issue with going over the indices during initialisation because from what I saw the information is pre-calculated and then serialised. So, here you create `policy -> indices`.

Then retrieving the data is just picking them up from the cache. Thoughts?